providers/enchant_nuspell.cpp

```diff
@@ -0,0 +1,441 @@
+/***************************************************************************/
+
+bool
+NuspellChecker::checkWord(const char *utf8Word, size_t len)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+
+	if (nuspell.spell(string(normalizedWord)))
+		return true;
+	else
+		return false;
+}
+
+char**
+NuspellChecker::suggestWord(const char* const utf8Word, size_t len, size_t *nsug)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+	auto suggestions = vector<string>();
+	nuspell.suggest(string(normalizedWord), suggestions);
+	if (suggestions.empty())
+		return nullptr;
+	*nsug = suggestions.size();
+	char **sug = g_new0 (char *, *nsug + 1);
+	size_t i = 0;
+	for (auto& suggest : suggestions) {
+		char *word = g_new0(char, suggest.size() + 1);
+		strcpy(word, suggest.c_str());
+		sug[i] = word;
+		i++;
+	}
+	return sug;
+}
+
+static void
+s_buildDictionaryDirs (vector<string> & dirs)
+{
+	dirs.clear ();
+
+	/* 1. personal overrides for Enchant
+	 *    ~/.config/enchant/nuspell
+	 */
+	gchar * tmp;
+	char * config_dir = enchant_get_user_config_dir ();
+	tmp = g_build_filename (config_dir, "nuspell", nullptr);
+	dirs.push_back (tmp);
+	free (config_dir);
+	g_free(tmp);
+
+	/* Dynamically retieved from Nuspell dictionary finder:
+	 * 2. personal overides for Hunspell
+	 *    ~/.local/share/hunspell
+	 * 3. system installed for Hunspell
+	 *    /usr/local/share/hunspell
+	 *    /usr/share/hunspell
+	 * 4. personal or system installations for
+	 *    Mozilla (such as Firefox and Thunderbird),
+	 *    LibreOffice and Apache OpenOffice
+	 */
+	auto nuspell_finder = nuspell::Finder::search_all_dirs_for_dicts();
+	for (auto& path : nuspell_finder.get_dir_paths()) {
+		if (path.compare(".") != 0)
+			dirs.push_back (path.c_str());
+	}
+
+	/* 5. system installations by Enchant
+	 *    /usr/local/share/enchant/nuspell
+	 *    /usr/share/enchant/nuspell
+	 */
+	char * enchant_prefix = enchant_get_prefix_dir();
+	if (enchant_prefix) {
+		tmp = g_build_filename(enchant_prefix, "share", "enchant", "nuspell", nullptr);
+		dirs.push_back (tmp);
+		g_free(enchant_prefix);
+		g_free(tmp);
+	}
+
+	/* Hunspell paths are used, therefore is ENCHANT_NUSPELL_DICT_DIR
+	 * irrelevant. Hence, the following paths are not to be considered:
+	 * /usr/local/share/nuspell and /usr/share/nuspell
+	 */
+}
+
+static void
+s_buildHashNames (vector<string> & names, const char * dict)
+{
+	names.clear ();
+
+	vector<string> dirs;
+	s_buildDictionaryDirs (dirs);
+
+	char *dict_dic = g_strconcat(dict, ".dic", nullptr);
+	for (size_t i = 0; i < dirs.size(); i++) {
+		char *tmp = g_build_filename (dirs[i].c_str(), dict_dic, nullptr);
+		names.push_back (tmp);
+		g_free (tmp);
+	}
+
+	g_free(dict_dic);
+}
+
+static const string
+s_correspondingAffFile(const string & dicFile)
+{
+	string aff = dicFile;
+	aff.replace(aff.end()-3, aff.end(), "aff");
```

Comment:
Sorry to be fussy about this, but please can you match and remove `".dic"` and then add `".aff"` here. The slow-down doesn't matter, and the resulting code is much clearer and more robust!

Reply:
This is identical code as Hunspell has See https://github.com/nuspell/enchant/commit/60077ae43b43691726dbae9d299dba1b5fb54deb#diff-b4a4b7f548e6522299f21f4f5632ff1cR176

Reply:
Oops. please see https://github.com/nuspell/enchant/commit/8844fa0464ad7076dc5f9849d726303409617bdc#diff-b4a4b7f548e6522299f21f4f5632ff1cR176